### ISSUE TITLE
Release 0.9.0

### ISSUE THREAD

**Issue by krakjoe** [krakjoe/parallel#31]
Upcoming release contains:

  - Channels: bi-directional communication between tasks and runtimes
  - Improvements to copying making it vastly more efficient
  - Improvements to runtime stack, making it more efficient
  - Tasks may use lexical scope
  - Events - an experimental, quite primitive event loop for reading/writing sets of channels and futures
  - Support for rethrowing exceptions uncaught in tasks
  - Simplified Runtime constructor
  - Remove Future::select (in favour of the superior events interface)
  - Task cancellation
  - Closure support

Will assign this to you @remicollet when I think ready ... ta ...

Anyone got any last minute thoughts before this next release happens ?

**Comment by krakjoe** [comment]
I've started work on documentation changes, they should be complete today ...

over to you @remicollet  :)

**Comment by krakjoe** [comment]
docs servers have updated, so I removed the stubs from the readme now to make it a bit shorter ... don't know if it's too short, maybe we could move the install.md stuff back into readme.md ?

**Comment by dktapps** [comment]
readme looks good to me.

**Comment by WyriHaximus** [comment]
Yup readme is fine 👍

**Comment by krakjoe** [comment]
The documentation is missing information about sharing closures ... because I was so excited to write it ... I'll do it today ...

I know we decided not to go for objects, but the ability to share a closure was much too appealing, they aren't really objects anyway, they just look like objects ... we have the ability to share functions so we should use it to the full, sharing closures allows some really cool stuff (and also golang allows it) ...

**Comment by dktapps** [comment]
how does the performance of closures compare to, for example, copying strings? right now in RakLib I use a string message channel to communicate messages between the main and child thread, and I'd love to switch it to closures if the performance of doing so is reasonable. 

in addition we had some problems with closures in pthreads once upon a time due to the caching - has this problem been accounted for with parallel?

**Comment by krakjoe** [comment]
> how does the performance of closures compare to, for example, copying strings?

That's really hard to answer in a simple way ... the structure of a closure is 10x the size of a 0 length zend_string, so the question of what takes up more memory depends on the length of the string ... the question of what is more efficient depends on what the string does or what you do to the string (do you eval() it on the recving side for example).

In parallel, copying works very differently to copying in pthreads. In parallel, the target mode is jit, jit'd code cannot be modified (nor does it need to be modified), nor can opcached code be modified. When parallel finds cached code (which may be jit'd) it handles it as core does, when parallel finds uncached code, it caches it in it's own simple opcode cache, and then handles the cached code as core does ...  cached code is copied using _one memcpy_ (and where statics are present a hash copy).

The closure itself is a couple of memcpy's when sending a closure, the same when receiving plus a std_object_init (something like `new` without constructor) ...

In other words, copying code in parallel is basically the only thing it needs to do, and it's getting quite good at it :)

**Comment by krakjoe** [comment]
Hold off release until I ping again ... I want to rewrite some of what I wrote in the last two days ...

**Comment by krakjoe** [comment]
@remicollet ping ... I've pushed the changes to docs that I wanted to get in and done my refactoring, everything looks pretty good to me, so over to you, for real this time :)

**Comment by remicollet** [comment]
@krakjoe sorry for the delay, looks like I missed the ping. Will start the release process today

**Comment by krakjoe** [comment]
@remicollet no problem, I rewrote everything a bunch of times while waiting anyway, so probably a good thing :)

ta

**Comment by remicollet** [comment]
@krakjoe 7.1 build is broken, working on it

**Comment by krakjoe** [comment]
7.1 is not supported, actually ZTS is not working there ... it's broken before 7.2, which I knew about (for years https://twitter.com/krakjoe/status/887743571515912196), but forgot when first writing parallel so supported it by accident ...

**Comment by remicollet** [comment]
OK, so will raise minimal dependencies...

**Comment by remicollet** [comment]
Released and announced
https://pecl.php.net/package-changelog.php?package=parallel

**Comment by krakjoe** [comment]
Thanks @remicollet  :)
